### PR TITLE
Feature: Snake Game Keybinds

### PR DESCRIPTION
**Pull request by j10a1n15**
## What
This Pull Request adds WASD Keybind support to the silly snake game.

## Changelog New Features
+ Added WASD keybinds to the Abiphone Snake Game. - j10a1n15

### CODE REVIEW

**Review by CalMWolfs — CHANGES_REQUESTED** [review]
could we instead use the player's minecraft keybinds instead of hard coding wasd

**Review by CalMWolfs — APPROVED** [review]
couldnt test but code looks fine

File: src/main/java/at/hannibal2/skyhanni/config/features/inventory/InventoryConfig.java [hunk 1]
```diff
@@ -304,6 +304,12 @@ public String toString() {
     @FeatureToggle
     public boolean hexAsColorInLore = true;
 
+    @Expose
+    @ConfigOption(name = "Snake Game Keybinds", desc = "Use WASD to move around in the AbiPhone snake game.")
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/config/features/inventory/InventoryConfig.java]: Hypixels spelling is with a lowercal `p`

Comment (hannibal002) [reply src/main/java/at/hannibal2/skyhanni/config/features/inventory/InventoryConfig.java]: maybe use `WASD-keys`

File: src/main/java/at/hannibal2/skyhanni/features/inventory/SnakeGame.kt [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+import at.hannibal2.skyhanni.utils.repopatterns.RepoPattern
+import net.minecraft.client.Minecraft
+import net.minecraft.client.gui.inventory.GuiChest
+import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
+import kotlin.time.Duration.Companion.milliseconds
+
+@SkyHanniModule
+object SnakeGame {
+
+    private val pattern by RepoPattern.pattern("snake.name", "Snake")
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/inventory/SnakeGame.kt]: better key

File: src/main/java/at/hannibal2/skyhanni/features/inventory/SnakeGame.kt [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    private val config get() = SkyHanniMod.feature.inventory
+    private var lastClick = SimpleTimeMark.farPast()
+
+    private val mcConfig
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/inventory/SnakeGame.kt]: remove line break

File: src/main/java/at/hannibal2/skyhanni/features/inventory/SnakeGame.kt [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        get() = Minecraft.getMinecraft().gameSettings
+
+    private val keys
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/inventory/SnakeGame.kt:27]: again remove line break

File: src/main/java/at/hannibal2/skyhanni/features/inventory/SnakeGame.kt [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        get() = mapOf(
+            mcConfig.keyBindLeft.keyCode to 50,
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/inventory/SnakeGame.kt]: if mcConfig is only used in here, use a local variable instead

File: src/main/java/at/hannibal2/skyhanni/features/inventory/SnakeGame.kt [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+            mcConfig.keyBindForward.keyCode to 51,
+            mcConfig.keyBindRight.keyCode to 52,
+            mcConfig.keyBindBack.keyCode to 53,
+        )
+
+    @SubscribeEvent
+    fun onGui(event: GuiKeyPressEvent) {
+        if (!isEnabled()) return
+        if (!pattern.matches(InventoryUtils.openInventoryName())) return
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/inventory/SnakeGame.kt]: check the inventory name on inventory open, not every tick. cache into a inInventory variable

File: src/main/java/at/hannibal2/skyhanni/features/inventory/SnakeGame.kt [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+        val chest = event.guiContainer as? GuiChest ?: return
+
+        if (lastClick.passedSince() < 100.milliseconds) return
+
+        for ((key, slot) in keys) {
+            if (key.isKeyHeld()) {
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/inventory/SnakeGame.kt]: early return

File: src/main/java/at/hannibal2/skyhanni/features/inventory/SnakeGame.kt [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+                event.cancel()
+
+                Minecraft.getMinecraft().playerController.windowClick(
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/inventory/SnakeGame.kt]: it starts getting time for a utils function for windowClick. maybe outside of this pr though

**Review by hannibal002 — CHANGES_REQUESTED** [review]
Also maybe make the keys cache on inventory open, so that we dont need to create the map every tick time.
I just assume the user cant change their keybinds without leaving the inventory

File: src/main/java/at/hannibal2/skyhanni/features/inventory/SnakeGame.kt (hunk 4 above)
Comment (j10a1n15) [reply src/main/java/at/hannibal2/skyhanni/features/inventory/SnakeGame.kt:27]: intellij auto formatting keeps setting a line break here

**Review by hannibal002 — APPROVED** [review]

**Pull request merged**